libraries/lib-music-information-retrieval/tests/MirTestUtils.h

```diff
@@ -0,0 +1,162 @@
+   // therefore if false positive rate has increased at some i, true positive
+   // rate has not decreased.
+   std::vector<double> truePositiveRates;
+   truePositiveRates.reserve(size);
+   std::vector<double> falsePositiveRates;
+   falsePositiveRates.reserve(size);
+   size_t numTruePositives = 0;
+   size_t numFalsePositives = 0;
+   for (const auto &result : results)
+   {
+      if (result.truth)
+         ++numTruePositives;
+      else
+         ++numFalsePositives;
+      truePositiveRates.push_back(
+         static_cast<double>(numTruePositives) / numPositives);
+      falsePositiveRates.push_back(
+         static_cast<double>(numFalsePositives) / numNegatives);
+   }
+
+   // Now find the area under the non-decreasing curve with FPR as x-axis,
+   // TPR as y, and i as a parameter.  (This curve is within a square with unit
+   // side.)
+   double auc = 0.;
+   for (size_t i = 0; i <= size; ++i)
+   {
+      const auto leftFpr = i == 0 ? 0. : falsePositiveRates[i - 1];
+      const auto rightFpr = i == size ? 1. : falsePositiveRates[i];
+      const auto leftTpr = i == 0 ? 0. : truePositiveRates[i - 1];
+      const auto rightTpr = i == size ? 1. : truePositiveRates[i];
+      const auto trapezoid = (rightTpr + leftTpr) * (rightFpr - leftFpr) / 2.;
+      assert(trapezoid >= 0); // See comments above
+      auc += trapezoid;
+   }
+
+   // Find the parameter at which the x coordinate exceeds the allowed FPR.
+   const auto it = std::upper_bound(
+      falsePositiveRates.begin(), falsePositiveRates.end(),
+      allowedFalsePositiveRate);
+
+   if (it == falsePositiveRates.end())
+      // All breakpoints satify the constraint. Return the least score.
+      return { auc, results.back().score };
+   else if (it == falsePositiveRates.begin())
+      // No breakpoint satisfies the constraint. Return the greatest score.
+      return { auc, results.front().score };
+
+   // For threshold, use the score halfway between the last breakpoint that
+   // satisfies the constraint and the first breakpoint that doesn't.
+   const auto index = it - falsePositiveRates.begin();
+   const auto threshold = (results[index - 1].score + results[index].score) / 2;
+
+   return { auc, threshold };
+}
+
+std::vector<std::string> GetWavFilesUnderDir(const char* dir);
+
+void ProgressBar(int width, int percent);
+
+template <typename T>
+void PrintPythonVector(
+   std::ofstream& ofs, const std::vector<T>& v, const char* name)
+{
+   ofs << name << " = [";
+   std::for_each(v.begin(), v.end(), [&](T x) { ofs << x << ","; });
+   ofs << "]\n";
+}
+
+struct OctaveError
+{
+   double factor;
+   double remainder;
+};
+
+OctaveError GetOctaveError(double expected, double actual);
+
+// A simple hashing that add the sum of all samples to the given hash.
+// Reproducible benchmarking must use the same input. We use this to make sure
+// that it does.
+template <int bufferSize = 1024>
+void UpdateHash(const MirAudioReader& source, float& hash)
```

Comment:
From @Paul-Licameli:
> Maybe write it in functional style, taking a value and returning a new value.
>
>I lift eyebrows to see a floating point type for a "hash" value. I looked for uses of this function. You are not really using it in a standard unordered container, which would require some quantization of the value to size_t. I think "checksum" better describes the use of the value.

Reply:
I agree, checksum is a better choice of name.
Also changing the signature to something more straightforward.